File: src/main/java/com/dg/watcher/watching/ApkWatching.java
package com.dg.watcher.watching;

import hudson.model.AbstractBuild;
import java.io.File;
import java.io.PrintStream;
import static com.dg.watcher.watching.retrieving.ApkRetrieving.retrieveApk;
import static com.dg.watcher.base.Const.BUILD_FORBIDDEN;
import static com.dg.watcher.base.Const.BUILD_ALLOWED;
import static com.dg.watcher.watching.saving.SizeSaving.loadApkSizes;
import static com.dg.watcher.watching.saving.SizeSaving.saveApkSize;
import static com.dg.watcher.watching.surveying.SizeSurveying.surveySizes;


public class ApkWatching {
    public static boolean watchApkSize(AbstractBuild<?, ?> build, PrintStream logger, float thresholdInMb, String customPathToApk) {
        File apk = retrieveApk(build, customPathToApk);

        if(apkFound(apk)) {
            saveApkSize(apk, build);

            return evaluateSize(build, logger, thresholdInMb);
        }
        else {
            return permitBuild(logger, thresholdInMb);
        }
    }

    private static boolean evaluateSize(AbstractBuild<?, ?> build, PrintStream logger, float thresholdInMb) {
        switch(surveySizes(loadApkSizes(build), thresholdInMb)) {
            case SIZE_THRESHOLD_EXCEEDED:
                return cancelBuild(logger, thresholdInMb);
            default:
                return permitBuild(logger, thresholdInMb);
        }
    }

    private static boolean apkFound(File apk) {
        return apk != null;
    }

    private static boolean cancelBuild(PrintStream logger, float thresholdInMb) {
        logger.println("Apk Watcher: Build Failed");
        logger.println("Apk Watcher: The size difference between the " +
                "last and the latest apk exceeded the specified threshold " +
                "of " + thresholdInMb + " megabyte.");

        return BUILD_FORBIDDEN;
    }

    private static boolean permitBuild(PrintStream logger, float thresholdInMb) {
        logger.println("Apk Watcher: Build Succeeded");
        logger.println("Apk Watcher: The size difference between the " +
                "last and the latest apk met the specified threshold " +
                "of " + thresholdInMb + " megabyte.");

        return BUILD_ALLOWED;
    }
}

File: src/main/java/com/dg/watcher/watching/ApkWatching.kt
package com.dg.watcher.watching

import com.dg.watcher.base.BUILD_ALLOWED
import com.dg.watcher.base.BUILD_FORBIDDEN
import com.dg.watcher.watching.retrieving.retrieveApk
import com.dg.watcher.watching.saving.loadApkSizes
import com.dg.watcher.watching.saving.saveApkSize
import com.dg.watcher.watching.surveying.SizeSurveyingResult.SIZE_THRESHOLD_EXCEEDED
import com.dg.watcher.watching.surveying.surveySizes
import hudson.model.AbstractBuild
import java.io.PrintStream


fun watchApkSize(build: AbstractBuild<*, *>, logger: PrintStream, thresholdInMb: Float, customPathToApk: String): Boolean {
    val apk = retrieveApk(build, customPathToApk)

    return if(apk != null) {
        saveApkSize(apk, build)

        evaluateSize(build, logger, thresholdInMb)
    }
    else {
        permitBuild(logger, thresholdInMb)
    }
}

private fun evaluateSize(build: AbstractBuild<*, *>, logger: PrintStream, thresholdInMb: Float) =
        when(surveySizes(loadApkSizes(build), thresholdInMb)) {
            SIZE_THRESHOLD_EXCEEDED -> cancelBuild(logger, thresholdInMb)

            else -> permitBuild(logger, thresholdInMb)
        }

private fun cancelBuild(logger: PrintStream, thresholdInMb: Float): Boolean {
    logger.println("Apk Watcher: Build Failed")
    logger.println("Apk Watcher: The size difference between the " +
            "last and the latest apk exceeded the specified threshold " +
            "of " + thresholdInMb + " megabyte.")

    return BUILD_FORBIDDEN
}

private fun permitBuild(logger: PrintStream, thresholdInMb: Float): Boolean {
    logger.println("Apk Watcher: Build Succeeded")
    logger.println("Apk Watcher: The size difference between the " +
            "last and the latest apk met the specified threshold " +
            "of " + thresholdInMb + " megabyte.")

    return BUILD_ALLOWED
}